contracts/clearing/TradingApp.sol

```diff
@@ -90,14 +32,55 @@ contract TradingApp is IForceMoveApp {
 
 		bytes memory candidateData = candidate.variablePart.appData;
 		// settlement
-		// TODO: unsure whether we should check the proof when consensus is reached
-		if (candTurnNum % 2 == 0 && proof.length == 0) {
+		if (candTurnNum % 2 == 0 && proof.length != 0) {
 			Consensus.requireConsensus(fixedPart, proof, candidate);
-			// NOTE: used just to check the data structure validity
-			ITradingStructs.Settlement memory _unused = abi.decode(
+			// Check the settlement data structure validity
+			ITradingTypes.Settlement memory _unused = abi.decode(
 				candidateData,
-				(ITradingStructs.Settlement)
+				(ITradingTypes.Settlement)
+			);
+
+			VariablePart memory prevProof = proof[proof.lenght - 1].variablePart;
+			ITradingTypes.Order memory prevOrder = abi.decode(
+				prevProof.appData,
+				(ITradingTypes.Order)
 			);
+			for (uint256 i = proof.lenght - 2; i >= 0; i--) {
```

Comment:
I think, it's better to verify Order-OrderResponse pairs, therefore increment the iterator by 2. This also means we can extract the `Order-OrderResponse` validation into a separate function to make requirements more explicit.

Reply:
There's a potential negative case that we miss by doing it this way: Nth Order, Nth OrderResponse, Nth+1 Order may have same order ID, and the updated algo would consider such sequence as valid

Reply:
Nice catch! Indeed, Signed Settlement is NOT tied to the proof in any way, meaning the proof can be tampered with, but the TradingApp will still successfully validate it. To mitigate this problem we can add a `ordersHash` field in the Settlement struct, which will be calculated as `keccak256(abi.encode(bytes32[]))` with elements being OrderIds in incremental order (from the first to the last order).